src/value.rs

```diff
@@ -0,0 +1,763 @@
+    }
+}
+
+impl<'frame, 'metadata, 'py, T> PyDeserializeValue<'frame, 'metadata, 'py> for Option<T>
+where
+    T: PyDeserializeValue<'frame, 'metadata, 'py>,
+{
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue, DeserializationError> {
+        match v {
+            Some(v) => T::deserialize_py(typ, Some(v), py),
+            None => Ok(PyDeserializedValue::new(py_none(py))),
+        }
+    }
+}
+
+pub struct PyDeserializedValue {
+    value: Py<PyAny>,
+}
+
+impl PyDeserializedValue {
+    fn new(value: Bound<PyAny>) -> Self {
+        Self {
+            value: value.into(),
+        }
+    }
+}
+
+impl<'py> IntoPyObject<'py> for PyDeserializedValue {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = Infallible;
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(self.value.into_bound(py))
+    }
+}
+
+impl<'py> IntoPyObject<'py> for &PyDeserializedValue {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = Infallible;
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(self.value.clone_ref(py).into_bound(py))
+    }
+}
+
+impl<'frame, 'metadata, 'py> PyDeserializeValue<'frame, 'metadata, 'py> for PyDeserializedValue {
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<Self, DeserializationError> {
+        match v {
+            None => Ok(PyDeserializedValue::new(py_none(py))),
+            Some(v) => {
+                let cql = deser_cql_py_value(py, typ, &mut v.as_slice())?;
+                Ok(PyDeserializedValue::new(cql))
+            }
+        }
+    }
+}
+
+fn deser_cql_py_value<'py, 'metadata, 'frame>(
+    py: Python<'py>,
+    typ: &'metadata ColumnType<'metadata>,
+    buf: &mut &'frame [u8],
+) -> Result<Bound<'py, PyAny>, DeserializationError> {
+    if buf.is_empty() {
+        match typ {
+            Native(NativeType::Ascii) | Native(NativeType::Blob) | Native(NativeType::Text) => {
+                // can't be empty
+            }
+            _ => return Ok(py_none(py)),
+        }
+    }
+
+    let val = Some(FrameSlice::new_borrowed(buf));
+
+    match typ {
+        Native(native_type) => match native_type {
+            NativeType::Counter => {
+                let v = Counter::deserialize(typ, val)?;
+                Ok(PyInt::new(py, v.0).into_any())
+            }
+            NativeType::Decimal => {
+                let dec = CqlDecimal::deserialize(typ, val)?;
+                let big: bigdecimal_04::BigDecimal = dec.into();
+                call_python_constructor(py, "decimal", "Decimal", (big.to_string(),), None)
+                    .map_err(DeserializationError::new)
+            }
+            NativeType::TinyInt => {
```

Comment:
Could you check what python-driver and scyllapy deserialize cql Decimal to? Do they also use `decimal.Decimal`?